File: token-validation-core/src/main/java/no/nav/security/token/support/core/configuration/MultiIssuerConfiguration.java
package no.nav.security.token.support.core.configuration;

import com.nimbusds.jose.util.ResourceRetriever;

import java.util.*;

public class MultiIssuerConfiguration {

    private final List<String> issuerShortNames = new ArrayList<>();
    private final ResourceRetriever resourceRetriever;

    private final Map<String, IssuerConfiguration> issuers = new HashMap<>();

    private final Map<String, IssuerProperties> issuerPropertiesMap;

    public MultiIssuerConfiguration(Map<String, IssuerProperties> issuerPropertiesMap) {
        this(issuerPropertiesMap, new ProxyAwareResourceRetriever());
    }

    public MultiIssuerConfiguration(Map<String, IssuerProperties> issuerPropertiesMap,
                                    ResourceRetriever resourceRetriever) {
        this.issuerPropertiesMap = issuerPropertiesMap;
        this.resourceRetriever = resourceRetriever;
        loadIssuerConfigurations();
    }

    public Map<String, IssuerConfiguration> getIssuers() {
        return issuers;
    }

    public Optional<IssuerConfiguration> getIssuer(String name) {
        return Optional.ofNullable(issuers.get(name));
    }

    public List<String> getIssuerShortNames() {
        return this.issuerShortNames;
    }

    public ResourceRetriever getResourceRetriever() {
            return resourceRetriever;
    }

    private void loadIssuerConfigurations() {

        issuerPropertiesMap.forEach((shortName, value) -> {
            issuerShortNames.add(shortName);
            var config = createIssuerConfiguration(shortName, value);
            issuers.put(shortName, config);
            issuers.put(config.getMetaData().getIssuer().toString(), config);
        });
    }

    private IssuerConfiguration createIssuerConfiguration(String shortName, IssuerProperties issuerProperties) {
        if (issuerProperties.getUsePlaintextForHttps() || issuerProperties.getProxyUrl() != null){
            var resourceRetrieverWithProxy = new ProxyAwareResourceRetriever(issuerProperties.getProxyUrl(), issuerProperties.getUsePlaintextForHttps());
            return new IssuerConfiguration(shortName, issuerProperties, resourceRetrieverWithProxy);
        }
        return new IssuerConfiguration(shortName, issuerProperties, resourceRetriever);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + " [issuerShortNames=" + issuerShortNames + ", resourceRetriever="
            + resourceRetriever + ", issuers=" + issuers + ", issuerPropertiesMap=" + issuerPropertiesMap + "]";
    }
}

File: token-validation-core/src/main/kotlin/no/nav/security/token/support/core/configuration/MultiIssuerConfiguration.kt
package no.nav.security.token.support.core.configuration

import com.nimbusds.jose.util.ResourceRetriever
import java.util.Optional
import kotlin.DeprecationLevel.WARNING

class MultiIssuerConfiguration @JvmOverloads constructor(private val properties : Map<String, IssuerProperties>, val retriever : ResourceRetriever = ProxyAwareResourceRetriever()) {

    val issuerShortNames = ArrayList<String>()

    val issuers : MutableMap<String, IssuerConfiguration> = HashMap()

    init {
        loadIssuerConfigurations()
    }
    @Deprecated(message ="Use of Optional not necessary",ReplaceWith("getIssuers.get(<issuer>)"), WARNING)
    fun getIssuer(name : String) = Optional.ofNullable(issuers[name])

    private fun loadIssuerConfigurations() =
        properties.forEach { (shortName, p) ->
             createIssuerConfiguration(shortName, p).run {
                issuerShortNames.add(shortName)
                issuers[shortName] = this
                issuers["${metadata.issuer}"] = this
            }
        }

    private fun createIssuerConfiguration(shortName : String, p : IssuerProperties) =
        if (p.usePlaintextForHttps || p.proxyUrl != null) {
            IssuerConfiguration(shortName, p, ProxyAwareResourceRetriever(p.proxyUrl, p.usePlaintextForHttps))
        }
        else IssuerConfiguration(shortName, p, retriever)

    override fun toString() = ("${javaClass.simpleName} [issuerShortNames=$issuerShortNames, resourceRetriever=$retriever, issuers=$issuers, issuerPropertiesMap=$properties]")
}